internal/rbd/controllerserver.go

```diff
@@ -294,27 +259,123 @@ func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
 	return &csi.CreateVolumeResponse{Volume: volume}, nil
 }
 
+func (cs *ControllerServer) checkFlatten(ctx context.Context, rbdVol *rbdVolume, cr *util.Credentials) error {
+	err := flattenRbdImage(ctx, rbdVol, cr)
+	if err != nil {
+		if _, ok := err.(ErrFlattenInProgress); ok {
+			return status.Error(codes.Aborted, err.Error())
+		}
+		if errDefer := deleteImage(ctx, rbdVol, cr); errDefer != nil {
+			klog.Errorf(util.Log(ctx, "failed to delete rbd image: %s with error: %v"), rbdVol, errDefer)
+			return status.Error(codes.Internal, err.Error())
+		}
+		errDefer := undoVolReservation(ctx, rbdVol, cr)
+		if errDefer != nil {
+			klog.Warningf(util.Log(ctx, "failed undoing reservation of volume: %s (%s)"), rbdVol.RequestName, errDefer)
+		}
+		return status.Error(codes.Internal, err.Error())
+	}
+	return nil
+}
+
+func (cs *ControllerServer) createVolumeFromSnapshot(ctx context.Context, cr *util.Credentials, rbdVol *rbdVolume, snapshotID string) error {
+	rbdSnap := &rbdSnapshot{}
+	if acquired := cs.SnapshotLocks.TryAcquire(rbdSnap.SnapID); !acquired {
```

Comment:
`rbdSnap.SnapID` is not set? The object just got created with empty values. Maybe you meant to use `snapshotID`?

Reply:
this should be fine, this function will be called after validating the snapshotID. this cannot be unset.

Reply:
No, I mean you create it on line 283 with `rbdSnap := &rbdSnapshot{}` and then use `rbdSnap.SnapID` on line 284. `rbdSnap.SnapID` will always be an empty string?